mbox: mbox_ti_secproxy: Driver optimization and bugfixes

Kronosblaster opened the pull request:
Improves the TI secure proxy mailbox driver with better error handling, per-channel interrupt support, and thread cleanup.

1: Fix thread verification
  - Replaced busy-wait loop with immediate error returns (-EINVAL, -EBUSY, -ENODATA)
  - Removed redundant message count checks
  - Simplified ISR error handling

  2: Support per-channel interrupts
  - Changed from single shared interrupt to per-channel interrupt array with individual ISR handlers
  - ISR now processes specific channels directly instead of polling all channels
  - Moved device tree configuration to core-specific files (am64x_r5f0_0.dtsi)
  - Added interrupt-names requirement to device tree binding

  3: Add thread flush on channel enable
  - Added secproxy_mailbox_flush_thread() to drain pending messages from RX thread
  - Automatically flushes thread when enabling channel to ensure clean state
  - Prevents stale messages from previous sessions

  4: Add SPDX License Identifier

Kronosblaster commented on drivers/mbox/mbox_ti_secproxy.c:
```diff
@@ -104,18 +107,13 @@ static inline int secproxy_verify_thread(struct secproxy_thread *spt, uint8_t di
+	if (dir == THREAD_IS_TX) {
+		if ((status & RT_THREAD_STATUS_CUR_CNT_MASK) == (SECPROXY_MAILBOX_NUM_MSGS)) {
+			return -EBUSY;
+		}
```
updated

Kronosblaster committed:
mbox: mbox_ti_secproxy: Flush rx thread on channel enable

Add secproxy_mailbox_flush_thread() to clear stale messages from
rx thread when enabling a channel. Read and discard all pending
messages to ensure clean thread state and prevent errors from
previous unclean state.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

natto1784 commented on drivers/mbox/mbox_ti_secproxy.c:
```diff
@@ -233,16 +220,9 @@ static int secproxy_mailbox_send(const struct device *dev, uint32_t channel,
 	spt.scfg = SEC_PROXY_THREAD(DEV_SCFG(dev), channel);
+	spt.channel = channel;
 
 	if (secproxy_verify_thread(&spt, THREAD_IS_TX)) {
```
better yet, return the value returned by `secproxy_verify_thread` if not zero instead of `-EBUSY`

Kronosblaster replied:
done

natto1784 commented on drivers/mbox/mbox_ti_secproxy.c:
```diff
@@ -359,25 +348,52 @@ static DEVICE_API(mbox, secproxy_mailbox_driver_api) = {
-	}                                                                                          \
+#define SECPROXY_THREAD_ISR(i, idx)                                                                \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(idx, DT_CAT(rx_, i)),                                     \
+		(	\
```
Move all backslashes in macros to column 99 to avoid formatting errors. clang-format won't fix a lot of these on its own.

natto1784 commented on drivers/mbox/mbox_ti_secproxy.c:
```diff
@@ -359,28 +351,54 @@ static DEVICE_API(mbox, secproxy_mailbox_driver_api) = {
+#define SECPROXY_IRQ_OR_ZERO(i, inst)                                                             \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(inst, DT_CAT(rx_, i)),                                   \
+		(DT_INST_IRQ_BY_NAME(inst, DT_CAT(rx_, i), irq)),                                 \
+		(0))
```
is it possible to have another placeholder - like a negative value? since 0 is a valid IRQ.

natto1784 commented on drivers/mbox/mbox_ti_secproxy.c:
```diff
@@ -38,15 +38,12 @@ LOG_MODULE_REGISTER(ti_secure_proxy);
 #define SEC_PROXY_DATA_START_OFFS 0x4
 #define SEC_PROXY_DATA_END_OFFS   0x3c
 
-#define SEC_PROXY_TIMEOUT_US 1000000
```
Hi, the rationale behind this change was that the timer is not initialized until `PRE_KERNEL_2`, hence it can cause issues to use busy_wait here. Alternatively, we can also add a check to see whether we are pre_kernel_* stages or not, but we concluded that this busy_wait is not necessary either way.

Kronosblaster committed:
mbox: mbox_ti_secproxy: Make verify thread function non-blocking

Remove busy wait from secproxy_verify_thread() to return immediately
with error status. If mailbox is full, waiting for message processing
is unnecessary and blocks execution until timeout, reporting a timeout
error message reducing debuggability.
The mailbox driver should only return the appropriate error code.
Any required retry logic can be implemented in the caller layer.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

Kronosblaster committed:
mbox: mbox_ti_secproxy: Add interrupt-names to DT binding

Add interrupt-names property to binding to enable per-channel
interrupt lookup via DT_INST_IRQ_BY_NAME(). Each channel's
interrupt is identified by name: "rx_0", "rx_1", "rx_2", etc.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

Kronosblaster committed:
mbox: mbox_ti_secproxy: Split AM64x R5F interrupts to per-core DT files

Split interrupt configuration from am64x_r5.dtsi into individual
R5F core files (am64x_r5f0_0.dtsi, am64x_r5f0_1.dtsi,
am64x_r5f1_0.dtsi, am64x_r5f1_1.dtsi) as each core uses
different mailbox channels and VIM interrupt line mappings.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

Kronosblaster committed:
mbox: mbox_ti_secproxy: Implement per-channel interrupt handling

Replace single shared interrupt with per-channel interrupt support.
Change .irq to .interrupts[] array in config, use LISTIFY macros
for ISR generation and connection, add channel_enable[] array for
per-channel state tracking. Each rx channel now has independent ISR.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

glneo commented on drivers/mbox/mbox_ti_secproxy.c:
```diff
@@ -295,6 +295,21 @@ static uint32_t secproxy_mailbox_max_channels_get(const struct device *dev)
+	spt.rt = SEC_PROXY_THREAD(DEV_RT(dev), channel);
+	spt.scfg = SEC_PROXY_THREAD(DEV_SCFG(dev), channel);
+
+	/* Drain all pending messages from the thread. Blocking call */
```
Blocking call where? Feels like this comment was copied from somewhere and doesn't apply here.

Kronosblaster replied:
You're right. Updated the comment to be clearer

Kronosblaster committed:
mbox: mbox_ti_secproxy: Remove redundant error checks

Remove duplicate error validation in ISR and send path that is
already performed by secproxy_verify_thread(). Remove error mask
checks, message queue fullness checks, and unused
SECPROXY_MAILBOX_NUM_MSGS macro.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

Kronosblaster committed:
mbox: mbox_ti_secproxy: Forward error codes in send path

Forward actual error codes from secproxy_verify_thread() instead
of masking all errors as -EBUSY. Now returns -EINVAL,
-EBUSY, or -ENODATA from secproxy_mailbox_send() to aid debugging.

Signed-off-by: Dave Joseph <d-joseph@ti.com>
